### PR TITLE
Check constraints on lifted operator types

### PR DESCRIPTION
**Pull request by 333fred**
We weren't validating constraints on lifted operators, which allowed us to generate types such as `Nullable<int*>` in operator binding without reporting any errors. Fixes https://github.com/dotnet/roslyn/issues/56646. The repro test case for that issue also hit an assert in `Binder.ValueCheck.cs`, where we were not expecting to run into a `BoundConditionalReceiver` node. I added it to the list of possible error scenario nodes.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
I think that under umbrella of this fix we should audit all the places where compiler constructs Nullable types. For example, stand alone conditional access comes to mind, etc. It also feels that, long term, it would be better to keep validity checks close to the code that constructs the types.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2). Tests in UserDefinedConversionTests are not looked at.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 1]
```diff
@@ -2784,6 +2784,7 @@ internal static uint GetValEscape(BoundExpression expr, uint scopeOfTheContainin
                 case BoundKind.AsOperator:
                 case BoundKind.AwaitExpression:
                 case BoundKind.ConditionalAccess:
+                case BoundKind.ConditionalReceiver:
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:2787]: I could add it there, but I don't have a test case to cover the change. If you want me to make the change without a test case I can.

[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:2787]: I added a sentence to the description.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [946068806](https://github.com/dotnet/roslyn/pull/57050#issuecomment-946068806) [](http://example.com/codeflow?ancestors=946068806,946068806)

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:2787]: > I could add it there, but I don't have a test case to cover the change. If you want me to make the change without a test case I can.

Let's at least open an issue to follow up, referencing the test scenario that prompted the change here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs [hunk 2]
```diff
@@ -1647,5 +1647,266 @@ private static C<object> M2()
...
+  .locals init (S? V_0)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""S?""
+  IL_0008:  ldloc.0
+  IL_0009:  stloc.0
+  IL_000a:  ldloca.s   V_0
+  IL_000c:  call       ""bool S?.HasValue.get""
+  IL_0011:  brtrue.s   IL_0017
+  IL_0013:  ldc.i4.0
+  IL_0014:  conv.u
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs:1900]: >conv.u [](http://example.com/codeflow?start=12&length=6)

Similar comment.

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs [hunk 3]
```diff
@@ -1647,5 +1647,266 @@ private static C<object> M2()
                 //         return (C<object>) M1();
                 Diagnostic(ErrorCode.ERR_NoExplicitConv, "(C<object>) M1()").WithArguments("void", "C<object>").WithLocation(9, 16));
         }
+
+        [Fact, WorkItem(56646, "https://github.com/dotnet/roslyn/issues/56646")]
+        public void LiftedConversion_InvalidTypeArgument01()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs:1652]: >LiftedConversion_InvalidTypeArgument01 [](http://example.com/codeflow?start=20&length=38)

It looks like verification should be disabled on desktop for test using fake special types.

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs [hunk 4]
```diff
@@ -1647,5 +1647,266 @@ private static C<object> M2()
...
+    public struct Void {}
+    public ref struct Int32 {}
+    public struct Nullable<T> where T : struct { public Nullable(T value) {} }
+    public struct Int64 {}
+    public class Attribute {}
+}
+";
+
+            var comp = CreateEmptyCompilation(code);
+            var verifier = CompileAndVerify(comp);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs:1713]: >CompileAndVerify [](http://example.com/codeflow?start=27&length=16)

It looks like verification should be disabled on desktop for test using fake special types.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:2787]: Filed [https://github.com/dotnet/roslyn/issues/57653](https://github.com/dotnet/roslyn/issues/57653)

File: src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs [hunk 5]
```diff
@@ -1647,5 +1647,266 @@ private static C<object> M2()
...
+  .locals init (S? V_0)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""S?""
+  IL_0008:  ldloc.0
+  IL_0009:  stloc.0
+  IL_000a:  ldloca.s   V_0
+  IL_000c:  call       ""bool S?.HasValue.get""
+  IL_0011:  brtrue.s   IL_0017
+  IL_0013:  ldc.i4.0
+  IL_0014:  conv.u
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs:1855]: No. This corresponds to `default(void*)`, which we emit as `ldc.i4.0` `conv.u`: [https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L2958-L2963](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L2958-L2963)

[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs:1855]: > No. This corresponds to default(void*), which we emit as ldc.i4.0 conv.u:

PEVerify complains about type mismatch across branches. Also, ```(void*)0``` is emitted as 
```
ldc.i4.0
conv.i
```  

What we emit for default is probably wrong.

[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UserDefinedConversionTests.cs:1855]: I've compared what we're emitting here vs what the native compiler was emitting:
```
{
  .entrypoint
  // Code size       44 (0x2c)
  .maxstack  1
  .locals init (valuetype [mscorlib]System.Nullable`1<valuetype S> V_0,
           void* V_1,
           valuetype [mscorlib]System.Nullable`1<valuetype S> V_2)
  IL_0000:  ldloca.s   V_0
  IL_0002:  initobj    valuetype [mscorlib]System.Nullable`1<valuetype S>
  IL_0008:  ldloc.0
  IL_0009:  stloc.2
  IL_000a:  ldloca.s   V_2
  IL_000c:  call       instance bool valuetype [mscorlib]System.Nullable`1<valuetype S>::get_HasValue()
  IL_0011:  brtrue.s   IL_0017
  IL_0013:  ldc.i4.0
  IL_0014:  conv.u
  IL_0015:  br.s       IL_0023
  IL_0017:  ldloca.s   V_2
  IL_0019:  call       instance !0 valuetype [mscorlib]System.Nullable`1<valuetype S>::GetValueOrDefault()
  IL_001e:  call       void* S::op_Implicit(valuetype S)
  IL_0023:  stloc.1
  IL_0024:  ldloc.1
  IL_0025:  conv.i4
  IL_0026:  call       void [mscorlib]System.Console::WriteLine(int32)
  IL_002b:  ret
}
```

The only differences are: the extra local for `S?`, and the store/load of the `void*` local, both of which we optimize out. I'm not certain why PEVerify is complaining for this code, but if we want to change what we emit for `default(void*)` that feels like it needs to be a separate change with the rammifications fully explored.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs [hunk 6]
```diff
@@ -11214,5 +11214,119 @@ static unsafe void M(dynamic d, int* p)
                 Diagnostic(ErrorCode.ERR_BadBinaryOps, "d += p").WithArguments("+=", "dynamic", "int*").WithLocation(5, 9)
                 );
         }
+
+        [Fact, WorkItem(56646, "https://github.com/dotnet/roslyn/issues/56646")]
+        public void LiftedUnaryOperator_InvalidTypeArgument01()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:11219]: Consider testing a conversion operator whose parameter type is `TypedReference`

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:11219]: How would you suggest getting a value to input into such a conversion?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:11219]: Oop, I think I see that this can't be done. Thanks.

**Pull request merged**